pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,476 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"path"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+const bitsUint32 = 32
+
+type bitArray struct {
+	data       []uint32
+	firstIndex int
+	size       uint32
+}
+
+func makeBitArray(size uint32) *bitArray {
+	m := bitArray{
+		data:       make([]uint32, (size+bitsUint32-1)/bitsUint32),
+		firstIndex: 0,
+		size:       size,
+	}
+	return &m
+}
+
+func (b *bitArray) set(index uint32) {
+	b.data[index/bitsUint32] |= (uint32(1) << (index % bitsUint32))
+}
+
+func (b *bitArray) isSet(index uint32) bool {
+	return (b.data[index/bitsUint32]>>(index%bitsUint32))&0x1 == 1
+}
+
+func (b *bitArray) findAndSetFirstZero() (uint32, error) {
+	for i := b.firstIndex; i < len(b.data); i++ {
+		if b.data[i] == 0xFFFFFFFF {
+			continue
+		}
+		for j := uint32(0); j < bitsUint32; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsUint32 + j
+				b.set(v)
+				b.firstIndex = int(i)
+				return v, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not find an empty slot")
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsUint32
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	b.data[i] &= ^(1 << (index % bitsUint32))
+}
+
+// length for the user namespace to create (65536).
+const userNsLength = math.MaxUint16 + 1
+
+// Limit the total number of pods using userns in this node to this value.
+// This is an alpha limitation that will probably be lifted later.
+const maxPods = 1024
+
+type usernsManager struct {
+	used     *bitArray
+	usedBy   map[string]uint32
+	removed  int
+	currPods int
+	kl       *Kubelet
+	sync.Mutex
+}
+
+// UserNamespace holds the configuration for the user namespace.
+type userNamespace struct {
+	// UIDs mappings for the user namespace.
+	UIDMappings []idMapping `json:"uidMappings"`
+	// GIDs mappings for the user namespace.
+	GIDMappings []idMapping `json:"gidMappings"`
+}
+
+// Pod user namespace mapping
+type idMapping struct {
+	// Required.
+	HostId uint32 `json:"hostId"`
+	// Required.
+	ContainerId uint32 `json:"containerId"`
+	// Required.
+	Length uint32 `json:"length"`
+}
+
+func MakeUserNsManager(kl *Kubelet) (*usernsManager, error) {
```

Comment:
I'd expect to see code like:
```go
if utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport)
```
here

Reply:
if we fail to create the object here then we need to make sure all callers check that `klet.usernsManager` is not nil before using it or a similar check.

Reply:
Could we return a no-op userns manager when the feature gate is disabled? We shouldn't add extra load to production clusters that aren't enabling the feature.

Reply:
I think it doesn't make sense to not allocate anything here. We _can_ probably allocate and don't do much more here, nor in release and other functions. Will update to that